pybandits/offline_policy_evaluator.py

```diff
@@ -0,0 +1,855 @@
+            prediction[:, action_index, :] = specific_action_prediction
+        return prediction
+
+
+class OfflinePolicyEvaluator(PyBanditsBaseModel, arbitrary_types_allowed=True):
+    """
+    Class to conduct OPE with multiple OPE estimatots
+    Based on obp package from Saito et al.:
+    Yuta Saito, Shunsuke Aihara, Megumi Matsutani, Yusuke Narita.
+    Open Bandit Dataset and Pipeline: Towards Realistic and Reproducible Off-Policy Evaluation
+    https://arxiv.org/abs/2008.07146
+    github:https://github.com/st-tech/zr-obp
+
+    Parameters
+    ----------
+    logged_data : pd.DataFrame
+        Logging data set
+    split_prop: Float01
+        Proportion of dataset used as training set
+    propensity_score_model_type: Literal["logreg", "gbm", "rf", "mlp", "batch_empirical", "empirical", "propensity_score"]
+        Method used to compute/estimate propensity score pi_b (propensity_score, logging / behavioral policy).
+    expected_reward_model_type: Literal["logreg", "gbm", "rf", "mlp"]
+        Method used to estimate expected reward for each action a in the training set.
+    n_trials : Optional[int]
+        Number of trials for the Optuna optimization process.
+    ope_metrics_list: Optional[List[BaseOffPolicyEstimator]]
+        List of OPE estimators used to evaluate the policy value of evaluation policy i.e. DirectMethod(),
+        InverseProbabilityWeighting(), DoublyRobust()
+    batch_feature: str
+        Column name for batch as available in logged_data
+    action_feature: str
+        Column name for action as available in logged_data
+    reward_feature: Union[str, List[str]]
+        Column name for reward as available in logged_data
+    contextual_features: Optional[List[str]]
+        Column names for contextual features as available in logged_data
+    cost_feature: Optional[str]
+        Column name for cost as available in logged_data; used for bandit with cost control
+    group_feature: Optional[str]
+        Column name for group definition feature as available in logged_data; available from simulated data
+        to define samples with similar contextual profile
+    true_reward_feature: Optional[Union[str, List[str]]]
+        Column names for reward proba distribution features as available in simulated logged_data. Used to compute ground truth
+    propensity_score_feature : Optional[str]
+        Column name for propensity score as available in logged_data; used for evaluation of the policy value
+    verbose : bool
+        Whether to log detailed information during the optimization process.
+    """
+
+    logged_data: pd.DataFrame
+    split_prop: Float01
+    propensity_score_model_type: Literal[
+        "logreg", "gbm", "rf", "mlp", "batch_empirical", "empirical", "propensity_score"
+    ]
+    expected_reward_model_type: Literal["logreg", "gbm", "rf", "mlp"]
+    scaler: Optional[Union[TransformerMixin, Dict[str, TransformerMixin]]] = None
```

Comment:
so now, any scaling strategy can be applied on contextual features?

Reply:
Indeed. You can also mix scalers, i.e. use different scalers for different contextual features.